### DISCUSSION TITLE
✨ Have sdk live as dependency

### DISCUSSION THREAD

**Discussion by Xiot** [dagger/dagger#12152] [Ideas]
### What are you trying to do?

Currently the typescript sdk installs itself as a folder within the project and has its own package.json and lock file.  
This makes things a little tricky to work around since there are now two separate places that try to manage the dependencies.  This is especially true since the dagger code references the sdk.

### Why is this important to you?

The 'normal' way for these types of dependencies are for it to be installed via a package manager and live within the `node_modules` folder.

### How are you currently working around this?

I haven't tried in a while, however the fact that the sdk had its own dependencies (and required it own package manager) than the main project left me with dependency errors that I wasn't able to resolve cleanly. 

This resulted me in not running dagger locally, and only running it through CI

**Comment by TomChv** [comment]
> Currently the typescript sdk installs itself as a folder within the project and has its own package.json and lock file.
This makes things a little tricky to work around since there are now two separate places that try to manage the dependencies. This is especially true since the dagger code references the sdk.

We installed the SDK as a local dependency because the SDK itself is bundled inside the Dagger binary.

We could do it differently but that means you cannot use the SDK as local dep... or we would need 2 different runtime, 1 local and 1 remote.

> I haven't tried in a while, however the fact that the sdk had its own dependencies (and required it own package manager) than the main project left me with dependency errors that I wasn't able to resolve cleanly.
>
> This resulted me in not running dagger locally, and only running it through CI

How is that so? What's your errors and which package manager are you using? 
When you say locally do you mean with `tsx xxx` or with `dagger call`? It should work with `dagger call` not matter the environment (CI, local, vps...)

**Comment by Xiot** [comment]
(Take this all with a grain of salt, since I haven't looked too deeply on how things are constructed internally)

> We installed the SDK as a local dependency because the SDK itself is bundled inside the 
Dagger binary.

That is the main part that I'm wondering about

> We could do it differently but that means you cannot use the SDK as local dep... or we would need 2 different runtime, 1 local and 1 remote

There are two unique parts.  One is the SDK itself, and the other is the code-gen part.
Those could be split up.  Some libraries that rely heavily on code-gen, like `prisma`, do this in order to keep a separation between the 'library' and the 'generated code'

> How is that so? What's your errors and which package manager are you using?
When you say locally do you mean with tsx xxx or with dagger call? It should work with dagger call not matter the environment (CI, local, vps...)

This was with `dagger call`.
I would have to take a look at it again, especially with the recent changes, but the issue was that the dependencies were conflicting since they were installed separately.

**Comment by TomChv** [comment]
> There are two unique parts. One is the SDK itself, and the other is the code-gen part.
Those could be split up. Some libraries that rely heavily on code-gen, like prisma, do this in order to keep a separation between the 'library' and the 'generated code'

That's actually a good idea, we might be able to split codegen & the SDK library, we would need 2 different important path though like

```typescript
import { Container } from "@dagger.io/sdk" // Generated code
import { object, func } from "@dagger.io/dagger" // Static library
```

Which means that sdk depends on `@dagger.io/dagger` but `@dagger.io/dagger` is a static library publised on npm on each release/push to main.
That also means we could publish our package to `jsr` and just create a specific imports header to supports deno.
That's super smart!

@helderco Do you have any opinion on this idea?

**Comment by Xiot** [comment]
The way that prisma does it (and there are other ways), is that it has its static library as `@prisma/client`.  Internally it references `.prisma/client` which is the pre-generated code that lives in `node_modules`.

This allows the user to import `@prisma/client` and have everything typed and working properly with their generated code.

The gritty details look something like
```
@prisma/client
  exports: 
    .: require('.prisma/client')
    ./runtime: // static code

.prisma/client
  imports:
    @prisma/client/runtime
```

The putting folders in `node_modules` causes some headaches, (also limits you to one usage in a monorepo), but there are other ways to get around that too

**Comment by Xiot** [comment]
Assumption Note:
This comment assumes that the code for a generated module is only needed in the engine.

Another thing that may work. Since you only really need the types for development, the 'client gen' could just spit out a `.d.ts` file that extends the dagger types (and creates new ones).

That file could then just be gitignored.

**Comment by helderco** [comment]
I’ve considered something similar for Python actually, but not exactly as suggested.

Let’s look at what makes up an SDK client library:
- The client bindings
- The client code that lives between the bindings and the communications with the API server, including a very simple connection logic based on only two env variables (session token and port)
- Code for supporting modules (introspection and registration, deserialization and serialization, validation and dynamic execution or dispatch)
- Automatic provisioning logic

The automatic provisioning is only useful outside of modules, and is a maintenance burden, especially since you can just use `dagger run` and take advantage of the TUI, but some people like it because it doesn’t require them to install the `dagger` CLI themselves.

There’s actually an [internal issue](https://linear.app/dagger/issue/DEV-3635/strip-the-sdk-directory-that-is-generated-for-modules) for stripping the automatic provisioning code in the SDK, and essentially trim it down to what’s necessary for modules. Basically make it easy for the engine to exclude this code when bundling for the runtime container. I actually started working on this a couple months ago for Python.

Then there’s the module support stuff which only makes sense in the context of modules and could be excluded from publishing to the public registries for example. It’s not even necessary to have locally since it’s only needed during runtime, and not for IDE autocompletions.

Now you’re just left with the generated client bindings and the underlying code that knows how to talk to the API. I actually think this is all that should be published to the public package registry.

For those that don’t know, Python was the second SDK to support modules and I initially intended to follow Go and generate something specific for modules, next to the client bindings, but I was pressed for time so I decided to just copy the whole SDK client library to the module and install it as a dependency with a local path source. 

However, this turned out to be an advantage in several different ways. It’s more easily isolated, it’s more flexible, and gives an easier path to support better integration with non module code, the host, etc. So we still have more experimentation we need to be free to make in this area.

There’s just a few wrinkles to fix, depending on SDK. We need to better decouple the SDK components, work on these integrations that we’re planning, and maybe at some point we’ll optimize further by leveraging the public package registry in modules.

But it’s too soon for that, and shouldn’t be required to make it work, only as an optimization. Imagine if you couldn’t test your non-Dagger TypeScript package locally while developing it. In this case it even hurts performance because you’re fetching the package from a remote server when right now it comes off the already running Dagger engine in your machine, with exactly the right compatible version between SDK and engine. So you’d only be “optimizing” for producing less clutter in your local filesystem, which only happens when you need IDE autocompletions, but possibly having a harder time keeping things appropriately in sync.

## To clarify

I think there’s maybe just a bit of confusion in how packages should be set up here. If any of the following points don’t work as described in the TypeScript SDK then it’s something that should be fixed, or that I’m unaware of, so point it out if that’s the case so we can discuss solutions.

For clarity:
- The generated `sdk` directory is only needed locally to support **IDE integration**. That’s it!
- You’re not required to install it locally for any other reason, it’s on `.gitignore` already, and can be safely deleted.
- Your module’s dependencies should at minimum list the local `sdk` path as a dependency, but it should install it into the module’s own `node_modules`, even if just a symlink (aka, editable dependency).
- There’s no reason to prefer having two copies of the SDK library (in `sdk` and inside `node_modules`) if a simple symlink will do.
- Tom tells me that there’s not even a need to install locally anyway since the IDE finds the necessary typing information from the path pointed at in `tsconfig.json`.
- With `sdk` as a dependency, the package manager **should not** install from the lock file in `sdk`, just like with published packages. I’d be surprised if a package manager does this on its own (I’d love to know about that), but even if it does, the SDK can just exclude that lock file from being bundled in the first place.
- To avoid discrepancies between the files that a module generates and those that are published, it’s possible for the TypeScript SDK to bundle a build of the client library, just like before it’s published to the registry. This would produce the same set of files when installed either way.
- I believe there’s a problem for monorepos in the JavaScript ecosystem when one package is a subdir of another, but I’d like to understand how TypeScript Dagger modules are problematic to use because of that.
- I’m convinced that the confusion around the lock file is from a previous bug in the TypeScript SDK runtime module, where at one point it was installing the SDK independently, before installing the module’s dependencies. I’m pretty sure that has been fixed though because I remember reviewing that. The runtime should only install the module’s own `package.json`, which will also install the SDK since it should be listed in there as a dependency.
- If the lock file in `sdk` is still being used somehow, then please file a bug report. That lock file should only be used when developing the client library in our repo’s `sdk/typescript`.

And bringing it home:
- Leveraging the public package registry is worse for performance, but there’s other ways to slim down what gets exported to the host. Only after all of that would it make sense to discuss offloading even more to the registry.
- I think it’s still worth it to decouple the client library components, especially the automatic provisioning logic, in order to strip that from the bundled SDK. Perhaps even not exporting the module support stuff to the host as well since it only makes sense to the runtime container and is not necessary for IDE integration.
- I’ve considered many times putting only the generated client bindings under `src/dagger` in the module, instead of vendoring the whole library. Basically what you’re suggesting, since it’s an itch I share! The problem is the change in imports, and we have a few things to figure out yet before cementing on something here.

**Comment by Xiot** [comment]
First, thanks for the great write up.  I'll go through this from top to bottom first, then have more freeform comments below.  I also want to point out that I only have knowledge/experience of using the dagger cli with modules, so I can't really comment on cases that fall outside of that

Initial reactions / comments

> Now you’re just left with the generated client bindings and the underlying code that knows how to talk to the API. I actually think this is all that should be published to the public package registry.

What are the `generated client bindings` here? I read that as it is the stuff that is generated from my 'client' module, although I'm most likely mistaken on this.

>  Imagine if you couldn’t test your non-Dagger TypeScript package locally while developing it

By test, do you mean running `dagger call` or more unit-testing style?  

>  So you’d only be “optimizing” for producing less clutter in your local filesystem, which only happens when you need IDE autocompletions, but possibly having a harder time keeping things appropriately in sync

I agree that keeping the versions in sync is a bit more of an issue.  You do have slightly the same issue now, as you can still have an old sdk installed locally that doesn't match the version of the dagger engine that you are running against.  

> The generated sdk directory is only needed locally to support IDE integration. That’s it!

Yup

> With sdk as a dependency, the package manager should not install from the lock file in sdk, just like with published packages.

The presence of the lock file causes the package managers to treat the folder differently (more later)

> at one point it was installing the SDK independently, before installing the module’s dependencies. 

I'm still getting errors running dagger locally, but I'll post a separate issue

> I’ve considered many times putting only the generated client bindings under src/dagger in the module, instead of vendoring the whole library. Basically what you’re suggesting, since it’s an itch I share!

Ya, I think we are pretty much on the same page here.

----

The main suggestion for splitting up the sdk, was to separating out the 'static' and 'dynamic' parts, and it was the static parts that I was originally suggesting that were to get published.  Anything that is generated, either from the local modules, or from dagger verse modules, should just live locally.

Since those dynamic parts are just for the IDEs, you could just build up a could `.d.ts` files.  That way you wouldn't even need much of anything installed.

And from the sounds of it, since the static parts are always just vendered from the local engine, you probably wouldn't need to put them in the folder structure either.

I never meant to imply that generated code should be published, as that causes a bunch of issues that you already called out

My main concern about having the `sdk` folder living within the normal project directory, is how it interacts with monorepos and node tooling in general.

In monorepo setups, you generally have a setting in the root `package.json` that tells it where to find the packages within the monorepo.  This can be configured to not include the `sdk` folder, but tooling generally expects `package.json` files that are outside of `node_modules` to be part of the monorepo.  `yarn` for instance, will throw an error if you try to `yarn install` in a folder that isn't in the includes `packages` property of the root `package.json`.  
There are valid reasons to have packages in a monorepo that aren't a part of it, and I have a couple in mine.  One of the ways to be able to configure, and install, its dependencies, is to include a `yarn.lock` file in that local package.

Note.. `yarn` v2+, by default (which is also what the sdk uses), uses `pnp` as its `node_linker`.  This approach doesn't even have a `node_modules` folder and it abstracts away the access to the dependencies in the `.pnp.js` file, which also acts as its lock file.

(side note, I currently have my dagger code outside of my monorepo in a separate repository. Part of this is due to technical limitations, but the main part is that I prefer having the build pipelines live separate from the rest of my code.  It allows me to iterate and change my build pipelines in a different cadence than the rest of the monorepo.  I can go over this in more depth later)

TLDR;
- I agree that if the sdk is only for the sdk, then we only need the type definitions locally.
- Having a separate `sdk` folder can be handled inside of a monorepo (but not one of the core packages) but it requires a specific configuration

**Comment by Xiot** [comment]
I just created https://github.com/dagger/dagger/issues/8722 for the issue where I can't run `dagger call` locally.  This is mainly caused by the fact that there are two separate package managers at play here.  This is a similar issue that would play out in a monorepo

**Comment by helderco** [comment]
> > Now you’re just left with the generated client bindings and the underlying code that knows how to talk to the API. I actually think this is all that should be published to the public package registry.
> 
> What are the `generated client bindings` here? I read that as it is the stuff that is generated from my 'client' module, although I'm most likely mistaken on this.

See this:
- https://github.com/dagger/dagger/issues/8424. 

Basically the “client bindings” is what codegen generates for the client, based on the API schema. It’s sometimes conflated with “module codegen” which does not only that, but also generates code and files specific for working with modules, like vendoring the client library in `sdk`.

When not using a module, codegen only generates client bindings for core API types. That’s what’s published in the public package registries today (along with the rest of the SDK’s client library).

When using a module, codegen also includes the client bindings for that module’s dependencies as well (this is the main reason for vendoring, but sync’d versions is another big one). And we have an outstanding issue to also be able to generate “self bindings” too, but that’s not possible today:
- https://github.com/dagger/dagger/issues/8030

One of the reasons for decoupling is we want to be able to generate client bindings from dependencies in a `dagger.json` for running directly on the host, instead of just the module environment:

- https://github.com/dagger/dagger/issues/5993
- https://github.com/dagger/dagger/issues/7707


> > Imagine if you couldn’t test your non-Dagger TypeScript package locally while developing it
> 
> By test, do you mean running `dagger call` or more unit-testing style?

I mean a normal TypeScript package that’s not a Dagger module. Using it locally during development, depending on use case. For a web app, or a CLI it could mean running directly, not just unit testing.

> I agree that keeping the versions in sync is a bit more of an issue. You do have slightly the same issue now, as you can still have an old sdk installed locally that doesn't match the version of the dagger engine that you are running against.

Not the same issue. If you have a different Dagger engine, then a `dagger develop` should update your local version of `sdk` automatically, without any other intervention from you. And remember, that local copy has no effect on whether or not the module is able to run. You can even delete it!

> > With sdk as a dependency, the package manager should not install from the lock file in sdk, just like with published packages.
> 
> The presence of the lock file causes the package managers to treat the folder differently (more later)

Yes, but only when installing from the directory that has the lock file directly, not from a dependency that has its own lock file. Do you see the difference?

In the first case you’re doing `npm install ./sdk`, but what I’m talking about is doing `npm install` when one of the dependencies is the local path `./sdk` (assuming there’s a lock file in `sdk`, which now shouldn’t exist).

> My main concern about having the `sdk` folder living within the normal project directory, is how it interacts with monorepos and node tooling in general.

That’s what I’m trying to understand, but I’m not following. Can you show a simple example? This would help me understand the constraints for:
- https://github.com/dagger/dagger/issues/8511

Why do you need to `npm install` the module locally? Is it just for IDE integration when developing the module or are you trying to import code from the module outside it in the monorepo? If it’s the former, isn’t `tsconfig.json` enough for that like Tom says? Is your only motivation to have a single `node_modules` directory in the monorepo for some reason? Does simply running `dagger init --sdk=typescript my-module` in an existing TypeScript/JavaScript project cause issues right off the bat? Does it affect the existing project or the Dagger module?

> In monorepo setups, you generally have a setting in the root `package.json` that tells it where to find the packages within the monorepo. This can be configured to not include the `sdk` folder

Do you mean the module’s folder? Or when you say “monorepo” here, you’re considering the module itself as the monorepo?

Because if you have a `<repo>/package.json`, you’ll also have a `<repo>/modules/my-module/package.json` and a `<repo>/modules/my-module/sdk/package.json`, so if the module is not the “monorepo” in this scenario then I’m thinking “what about the module’s `package.json`” 😅

You shouldn’t have a `<repo>/modules/my-module/sdk/node_modules/` or install that directory directly. It’s only there as a dependency source.

> but tooling generally expects `package.json` files that are outside of `node_modules` to be part of the monorepo. 

You mean down the file tree, not up, right? For example, having a `<repo>/node_modules/`, you’re saying tooling expects `<repo>/modules/my-module/package.json` to be a part of `<repo>/node_modules/` instead of having its own `<repo>/modules/my-module/node_modules/`?

> `yarn` for instance, will throw an error if you try to `yarn install` in a folder that isn't in the includes `packages` property of the root `package.json`. There are valid reasons to have packages in a monorepo that aren't a part of it, and I have a couple in mine. One of the ways to be able to configure, and install, its dependencies, is to include a `yarn.lock` file in that local package.

Is that really the best way to say you don’t want sub-package `<repo>/modules/my-module/` to be a part of the `<repo>/` monorepo? 😕 Since the TypeScript SDK already manages a lock file using your chosen package manager, doesn’t that solve the issue?

> Note.. `yarn` v2+, by default (which is also what the sdk uses), uses `pnp` as its `node_linker`. This approach doesn't even have a `node_modules` folder and it abstracts away the access to the dependencies in the `.pnp.js` file, which also acts as its lock file.

Not sure I follow. Is it good or bad? If there’s things we can change in the SDK to make life easier, we’re listening. 😇

> (side note, I currently have my dagger code outside of my monorepo in a separate repository. Part of this is due to technical limitations, but the main part is that I prefer having the build pipelines live separate from the rest of my code. It allows me to iterate and change my build pipelines in a different cadence than the rest of the monorepo. I can go over this in more depth later)

That was our preference initially 😄, but a lot of people preferred co-hosting the module next to the code that it manages. These days there’s additional benefits due to the introduction of default path for directories, but whenever possible we like to encourage this pattern actually. We’re still trying to figure out the best patterns that works best for both use cases.

**Comment by Xiot** [comment]
> One of the reasons for decoupling is we want to be able to generate client bindings from dependencies in a dagger.json for running directly on the host, instead of just the module environment

Ya, I'm only really considering the module case, so a bunch of my assumptions is probably incorrect or at the very lease imcomplete.


> Not the same issue. If you have a different Dagger engine, then a dagger develop should update your local version of sdk automatically, without any other intervention from you. And remember, that local copy has no effect on whether or not the module is able to run. You can even delete it!

It is the same issue in that I can develop the module using one set of typings, but those may change depending on the version of the engine that is running it.  
I ran into a similar issue with `Container.withNewFile`.  At one point, it only took 2 arguments, and the contents were passed in as a property on the options.  We upgraded the version of the engine, but I didn't change the version of the module.  Surprisingly the code didn't fail at runtime, but my IDE was showing errors when I did eventually upgrade the sdk locally.

> > Note.. yarn v2+, by default (which is also what the sdk uses), uses pnp as its node_linker. This approach doesn't even have a node_modules folder and it abstracts away the access to the dependencies in the .pnp.js file, which also acts as its lock file.

> Not sure I follow. Is it good or bad?

It was neither good, nor bad.  Just an explanation 


> but a lot of people preferred co-hosting the module next to the code that it manages.

I've been getting pressure internally to move the dagger code into the monorepo, so I understand the desire to keep everything together


For most of the other comments, it was mostly based on the fact that I was on an old version, which were causing things to break.  Now thats been fixed, I'm not really worried about it as much.

**Comment by TomChv** [comment]
Related to that issue: https://github.com/dagger/dagger/pull/9531.
If extended to module's client gen, this would use the published library :)

**Comment by TomChv** [comment]
Still a draft but it will solve that issue: https://github.com/dagger/dagger/pull/10846
